### PR TITLE
Add contributing section on code review

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

The text here is looking predominantly at what should (and shouldn't) be the subject of a code review, and how both the reviewer and PR author should try and behave to make things nicer and more useful for everyone.

### Details and comments

Close #6641 (replaces it).  I've tried to incorporate much of what was already in that PR and suggested on that PR, though I cut down some of the "internal" details about who constitutes a maintainer, etc, which might be a bit more appropriate for a governance document.

cc: @1ucian0, @javabster - probably you two want to look at this with your community / dev-rel hats on.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -66,6 +67,111 @@ please ensure that:
    deprecation, removal etc) that you have added a reno release note for that
    change and that the PR is tagged for the changelog.
 
+
+### Code review
+
+All code merged to Qiskit, even from maintainers, goes through a code-review
+process after a pull request is made.  We only have a small number of
+maintainers who can authorise a merge, and it's useful for everyone if the
```
Comment (mtreinish) [inline CONTRIBUTING.md:75]:
```suggestion
maintainers who can authorize a merge, and it's useful for everyone if the
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -66,6 +67,111 @@ please ensure that:
...
+community can take part in helping make sure pull requests are technically
+correct, well tested, and easy to review.
```
Comment (mtreinish) [inline CONTRIBUTING.md:77]: One thing I did in #6641 is to explicitly call out the value of people doing code review and also helping to reduce the maintainer burden. I think it would be good to reinforce this a bit more here. At least for me a big motivation in the original PR was to try and more strongly reinforce the value of having a culture of mutual review regardless of whether you empowered to merge code.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -66,6 +67,111 @@ please ensure that:
...
+  returning `None`?", rather than "you should raise an exception here".
+* Try to make each round of review thorough.  Don't add one or two comments on
+  one file, then come back a day later and add a couple of other unrelated
+  comments on a different file, and so on.  Try to review the whole PR
+  thoroughly in one go; it's easier to catch bugs like this, and less
+  frustrating to be on the other side of. If that's too much for you, that might
+  mean that the PR should be split into smaller chunks.
+* Try to keep the number of comments reasonable.  This depends on the size of
+  the PR, but remember that there's somebody who'll read all your comments, and
+  it feels rubbish if you get a PR back and it's got 30 comments on from a
```
Comment (mtreinish) [inline CONTRIBUTING.md:121]:
```suggestion
  it can be demoralizing if you get a PR back and it's got 30 comments on from a
```

**Review by mtreinish — COMMENTED** [review]
Thanks for reviving this section in the CONTRIBUTING guide I think it'll be a strong addition. Just a few small suggestions inline

File: CONTRIBUTING.md (hunk 1 above)
Comment (jakelishman) [reply CONTRIBUTING.md:75]: is this really the only British-ism?

File: CONTRIBUTING.md (hunk 2 above)
Comment (jakelishman) [reply CONTRIBUTING.md]: How's ef1c3441a?

Comment (mtreinish) [reply CONTRIBUTING.md]: Thanks, yeah that looks better to me.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+* Make concrete suggestions when you think something should be changed, but
+  remember that the author might have already thought about it and have a
+  reason.  Try "what do you think about us raising a `TypeError` here instead of
+  returning `None`?", rather than "you should raise an exception here".
+* Try to make each round of review thorough.  Don't add one or two comments on
+  one file, then come back a day later and add a couple of other unrelated
+  comments on a different file, and so on.  Try to review the whole PR
+  thoroughly in one go; it's easier to catch bugs like this, and less
+  frustrating to be on the other side of. If that's too much for you, that might
+  mean that the PR should be split into smaller chunks.
```
Comment (kevinhartman) [inline CONTRIBUTING.md:119]:
```suggestion
  frustrating to be on the other side of. If that's too much for you, that might
  mean that the PR should be split into smaller chunks.
```

This would be on the PR author, right? I think we ought to drop that last sentence, or suggest that they ask the PR author to make the PR more focused.

**Review by kevinhartman — COMMENTED** [review]
LGTM.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+All code merged to Qiskit, even from maintainers, goes through a code-review
+process after a pull request is made.  There are a small number of
+maintainers who can authorize a final merge, but code review is everyone
+working together to make Qiskit better.  You can review code even if you
+are not a maintainer, which helps make sure pull requests are technically
+correct, well tested, and easier in final review.
+
+The code-review process is a normal part of software development, and nothing to
+be scared of; for very easy changes it can be as simple as a maintainer saying
+"looks good to me!" and merging the PR.  For more complex changes, it's often a
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md:82]: As far as I can recall, I've never used that acronym (most just out of stubbornness at this point lol).

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+
+Remember that the PR author is a human, not just a screen name!  It's ok to ask
+questions about the code, but don't be mean or rude about it even if you don't
+like it.
+
+#### What to focus on in review
+
+* Is everything in [the PR checklist](#pull-request-checklist) done?
+* Are any new public APIs easy to use, well documented, and in keeping with
+  other parts of the Qiskit API?
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md]: Done in 2971995d8.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+"looks good to me!" and merging the PR.  For more complex changes, it's often a
+back-and-forth where the reviewer may ask a couple of questions about why things
+were done a particular way, and make suggestions for improvement.  You don't
+need to do everything suggested if you've got good reasons to disagree, but
+communicate that clearly and politely.
+
+If you're struggling with code review or a PR on Qiskit, you can ask for help in
+the `#qiskit-pr-help` channel on [the public Qiskit Slack](https://qisk.it/join-slack).
+
+Remember that the PR author is a human, not just a screen name!  It's ok to ask
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md]: I said "username", but otherwise done in 2971995d8.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+questions about the code, but don't be mean or rude about it even if you don't
+like it.
+
+#### What to focus on in review
+
+* Is everything in [the PR checklist](#pull-request-checklist) done?
+* Are any new public APIs easy to use, well documented, and in keeping with
+  other parts of the Qiskit API?
+* Do any changes to the code have knock-on effects for other parts of Qiskit
+  that may be using them, or do the changes imply changes to the assumptions in
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md]: Done in 2971995d8.

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinhartman) [reply CONTRIBUTING.md:82]: Heh, well. Matt and I use it at least :3

File: CONTRIBUTING.md (hunk 4 above)
Comment (kevinhartman) [reply CONTRIBUTING.md:119]: That'd be reasonable for them to infer. IMO though, it's a bit hand-wavy as it is, since we're putting stock on the competence of the reader to assess whether or not their review fatigue comes from a poorly factored PR.

Comment (jakelishman) [reply CONTRIBUTING.md]: Fair enough - I tweaked the wording in d62c1eea2, which I think should help?

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply CONTRIBUTING.md:82]: I do use it too.. so. Commiting :)

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -66,6 +67,112 @@ please ensure that:
    deprecation, removal etc) that you have added a reno release note for that
    change and that the PR is tagged for the changelog.
 
+
+### Code review
+
+All code merged to Qiskit, even from maintainers, goes through a code-review
+process after a pull request is made.  There are a small number of
+maintainers who can authorize a final merge, but code review is everyone
```
Comment (javabster) [inline CONTRIBUTING.md:75]:
```suggestion
maintainers who can authorize a final merge, but code review involves everyone
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+
+#### What to focus on in review
+
+* Is everything in [the PR checklist](#pull-request-checklist) done?
+* Are any new public APIs easy to use, well documented, and consistent with
+  other parts of the Qiskit API?
+* Do any changes to the code have knock-on effects for other parts of Qiskit
+  that may be using them, or do they imply changes to the assumptions in
+  our data structures?
+* Are there any edge cases you can think of that the code might not handle well?
```
Comment (javabster) [inline CONTRIBUTING.md:103]:
```suggestion
* Are there any edge cases you can think of that the code might not handle well? And are extra tests needed to cover these edge cases?
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+  frustrating to be on the other side of. If that's too much for you, consider
+  asking if the PR could be split into smaller independent chunks.
+* Try to keep the number of comments reasonable.  This depends on the size of
+  the PR, but remember that there's somebody who'll read all your comments, and
+  it can be demoralizing if you get a PR back and it's got 30 comments on from a
+  50-line change.  If you feel like you're putting too many comments on,
+  consider if you could group several of them into one theme, and ask them as a
+  more detailed question with a focus on only one part of the code.  Try not to
+  comment the same thing in many places.
+* Try to avoid saying "you do" in review comments, and instead try to
```
Comment (javabster) [inline CONTRIBUTING.md:127]: I'm a bit confused about this you do / we do comment, I'm not sure what exactly you're trying to say 😅 Maybe add an example?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+  say things like "we do" even when talking about new code. It's not a big
+  change, yet it helps to make us think about Qiskit's code as something that we
+  all own and care about, and that we're all working together to make it better.
+
+#### Responding to review comments
```
Comment (javabster) [inline CONTRIBUTING.md:132]: do you think it's worth adding a point to this section about respecting the expertise of core maintainers? Something about how it's ok to push back or question a review but at the end of the day if a maintainer is asking for a change you should respect and action their request if you want them to merge your work.

**Review by javabster — COMMENTED** [review]
Overall looking good and very detailed!

File: CONTRIBUTING.md (hunk 12 above)
Comment (jakelishman) [reply CONTRIBUTING.md:132]: I considered this when writing, and wasn't super sure whether I should say something like it or not.  I don't want to give the impression that maintainers are infallible, and we really would want maintainers to be able to back up change requests with convincing arguments when required (otherwise are we sure they're not pure opinion, which we want to avoid commenting on?), but I also do think that at some point yeah, the maintainer is the final say even if you don't fully buy their argument.

I can expand saying "the maintainer has the final say" (in nicer language), if that's something we all agree we should have?

File: CONTRIBUTING.md (hunk 9 above)
Comment (jakelishman) [reply CONTRIBUTING.md:75]: Personally I think "is" is fine here, but I don't mind.  Changed in 724d18de7.

File: CONTRIBUTING.md (hunk 10 above)
Comment (jakelishman) [reply CONTRIBUTING.md:104]: Done similar in 724d18de7.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -66,6 +67,112 @@ please ensure that:
...
+  else:
+      my_first_variable = 456
+      my_second_variable = 123
+  ```
+  there's no need to suggest changes like
+  ```python
+  my_first_variable = 123 if my_condition else 456
+  my_second_variable = 456 if my_condition else 123
+  ```
+  Both are perfectly legible, and focussing on small details like this is
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md]: Ah, another British-ism.  Americanised in 724d18de7.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -66,6 +67,115 @@ please ensure that:
    deprecation, removal etc) that you have added a reno release note for that
    change and that the PR is tagged for the changelog.
 
+
+### Code review
+
+All code merged to Qiskit, even from maintainers, goes through a code-review
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md:73]: At least in British English, the hyphen is correct: "code-review" is a compound adjective modifying "process" here.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -66,6 +67,115 @@ please ensure that:
...
+The code-review process is a normal part of software development, and nothing to
+be scared of; for very easy changes it can be as simple as a maintainer saying
+"looks good to me!" (or in short-form, "LGTM!") and merging the PR.  For more complex changes, it's often a
+back-and-forth where the reviewer may ask a couple of questions about why things
+were done a particular way, and make suggestions for improvement.  You don't
+need to do everything suggested if you've got good reasons to disagree, but
+communicate that clearly and politely.
+
+If you're struggling with code review or a PR on Qiskit, you can ask for help in
+the `#qiskit-pr-help` channel on [the public Qiskit Slack](https://qisk.it/join-slack).
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md:89]: The word "channel" is a little before the link - the whole sentence reads as "the ... channel on the public Qiskit Slack", so it'd be wrong to add an extra "channel".

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -66,6 +67,115 @@ please ensure that:
...
+### Code review
+
+All code merged to Qiskit, even from maintainers, goes through a code-review
+process after a pull request is made.  There are a small number of
+maintainers who can authorize a final merge, but code review involves everyone
+working together to make Qiskit better.  You can review code even if you
+are not a maintainer, which helps make sure pull requests are technically
+correct, well tested, and easier to tackle in their final maintainer review.
+
+The code-review process is a normal part of software development, and nothing to
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply CONTRIBUTING.md:80]: Same here - the hyphen is correct in British English, and I think American too.

File: CONTRIBUTING.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (jlapeyre) [reply CONTRIBUTING.md:254]: Same in American English

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -192,40 +196,38 @@ please ensure that:
    If your code fails the local style checks (specifically the black
    or Rust code formatting check) you can use `tox -eblack` and
```
Comment (1ucian0) [inline CONTRIBUTING.md:197]: Links, in case contributors are not aware of them:

```suggestion
   If your code fails the local style checks (specifically the [black](https://black.readthedocs.io/en/stable/)
   or [Rust code formatting check](https://rust-lang.github.io/rustfmt/)) you can use `tox -eblack` and
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -192,40 +196,38 @@ please ensure that:
...
+
 2. The documentation has been updated accordingly. In particular, if a
    function or class has been modified during the PR, please update the
    *docstring* accordingly.
 
    If your pull request is adding a new class, function, or module that is
    intended to be user facing ensure that you've also added those to a
    documentation `autosummary` index to include it in the api documentation.
+
 3. If you are of the opinion that the modifications you made warrant additional tests,
```
Comment (1ucian0) [inline CONTRIBUTING.md:208]: Less "optional" language?
```suggestion
3. If the modifications you made warrant additional tests,
```

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -192,40 +196,38 @@ please ensure that:
...
-   feel free to include them
+   feel free to include them.
```
Comment (1ucian0) [inline CONTRIBUTING.md:209]: Less "optional" language?

```suggestion
   please include include them.
```

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -192,40 +196,38 @@ please ensure that:
...
+
 4. Ensure that if your change has an end user facing impact (new feature,
-   deprecation, removal etc) that you have added a reno release note for that
+   deprecation, removal etc) that you have added a `reno` release note for that
```
Comment (1ucian0) [inline CONTRIBUTING.md:212]: trying to give examples beyond API changes

```suggestion
   deprecation, removal, bugfix, etc) that you have added a `reno` release note for that
```

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -192,40 +196,38 @@ please ensure that:
...
    change and that the PR is tagged for the changelog.
```
Comment (1ucian0) [inline CONTRIBUTING.md:213]: Give a bit more importance to the PR changelog labeling
```suggestion
   change.
   
   5. Make sure that the PR gets tagged with `Changelog: X` kind of label. If you are not sure which label you should use, leave it as a comment in your PR. 
```

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -192,40 +196,38 @@ please ensure that:
...
-are easier for maintainers to review and more likely to get merged in a timely manner. Please also make
-sure to always be kind and respectful in your interactions with maintainers and other contributors, you can read
-[the Qiskit Code of Conduct](https://github.com/Qiskit/qiskit/blob/main/CODE_OF_CONDUCT.md).
+   The PR title will become the summary line of the commit, which appears in `git log`.
+   For example, "Fixes Issue 1234" is a bad title, and "Fix `ApplyLayout` with
+   empty layouts" is good.
+
+7. If the PR addresses an open issue the PR description includes the `fixes #issue-number`
+   syntax to link the PR to that issue (**you must use the exact phrasing in order for GitHub
+   to automatically close the issue when the PR merges**)
```
Comment (1ucian0) [inline CONTRIBUTING.md:230]:
```suggestion
   syntax to link the PR to that issue (**you must use [specific phrasing](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) in order for GitHub
   to automatically close the issue when the PR merges**)
```

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
 Remember that spamming issues or pull requests with AI-generated comments is prohibited under the [Qiskit Code of Conduct](https://qisk.it/coc).
 
 
+### Code review
+
+All code merged to Qiskit, even from maintainers, goes through a code-review
+process after a pull request is made.  There are a small number of
+maintainers who can authorize a final merge, but code review involves everyone
+working together to make Qiskit better.  You can review code even if you
```
Comment (1ucian0) [inline CONTRIBUTING.md:257]:
```suggestion
maintainers who can authorize a final merge, but **code review involves everyone
working together to make Qiskit better**.  You can review code even if you
```

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+are not a maintainer, which helps make sure pull requests are technically
+correct, well tested, and easier to tackle in their final maintainer review.
+
+The code-review process is a normal part of software development, and nothing to
+be scared of; for very easy changes it can be as simple as a maintainer saying
+"looks good to me!" (or in short, "LGTM!") and merging the PR.  For more complex changes, it's often a
```
Comment (1ucian0) [inline CONTRIBUTING.md:263]: Nothing scares me as somebody saying "nothing to be scared of" `:P`
```suggestion
The code review process is a normal part of software development and is often simpler than expected. For very small changes, it can be as straightforward as a maintainer saying "looks good to me!" (or in short, "LGTM!") and merging the PR. For more complex changes, it's often a
```

File: CONTRIBUTING.md [hunk 25]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+#### What to focus on in review
+
+* Is everything in [the PR checklist](#pull-request-checklist) done?
+
+* Are any new public APIs easy to use, well documented, and consistent with
+  other parts of the Qiskit API?
+
+* Do any changes to the code have knock-on effects for other parts of Qiskit
+  that may be using them, or do they imply changes to the assumptions in
+  our data structures?
```
Comment (1ucian0) [inline CONTRIBUTING.md:286]:
```suggestion
  existing data structures?
```

File: CONTRIBUTING.md [hunk 26]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+
+* Are there any edge cases you can think of that the code might not handle well?
+  Could the PR benefit from extra tests to cover these, or to verify other edge
+  cases that it *does* handle successfully?
+
+* Is the code reasonably easy for you to understand?  This particular point is
+  tricky; the more you review code, the easier it will be for you to understand
+  other code, so don't worry about this as much if you're getting started.
+
+* If this PR is a bugfix, is it suitable for backport?  If not, could the PR be
```
Comment (1ucian0) [inline CONTRIBUTING.md:296]: Maybe it helps to understand this point (and even the implication of backporting) if:
```suggestion
* If this PR is a bugfix, is it suitable for backport to supported stable versions?  If not, could the PR be
```

File: CONTRIBUTING.md [hunk 27]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+  split into a "simple" bugfix that is suitable for backport and a follow-on
+  improvement?  (Not all bugfixes *must* be backported.)
+
+#### Writing review comments
+
+* Make concrete suggestions when you think something should be changed, but
+  remember that the author might have already thought about it and have a
+  reason.  Try "What do you think about us raising a `TypeError` here instead of
```
Comment (1ucian0) [inline CONTRIBUTING.md:304]: sometimes, the use of "us" feels like a counter to "them". Probably the distinction is pointless in this sentence anyway:
```suggestion
  reason.  For example, try "What do you think about raising a `TypeError` here instead of
```

File: CONTRIBUTING.md [hunk 28]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+  change, yet it helps to make us think about Qiskit's code as something that we
+  all own and care about, and that we're all working together to make it better.
+
+#### Responding to review comments
+
+* Ask questions if you don't understand what a reviewer is saying, or if you're
+  not certain whether they're suggesting changes.
+
+* Feel free to respond to suggestions or questions with your reasoning for doing
+  things a different way, if you don't fully agree with the review comment.
```
Comment (1ucian0) [inline CONTRIBUTING.md:333]: "feel free" sounds _very_ optional to me.

```suggestion
* Consider responding to suggestions or questions with your reasoning for doing
  things a different way, if you don't fully agree with the review comment.
```

File: CONTRIBUTING.md [hunk 29]
```diff
@@ -247,6 +249,183 @@ Submissions that appear unreviewed or copied directly from an AI tool without pr
...
+  ```
+  @Mergifyio backport <branch> <branch2> ...
+  ```
+
+  You can have as many branches as necessary.  It usually only necessary to do
+  this to support old major branches.
+
+* Any issues fixed by the PR have their own "Fix #<num>" line in the author's PR
+  comment.  If you are empowered to merge PRs, you should be empowered to edit
+  the author's comment to add these, if necessary.
```
Comment (1ucian0) [inline CONTRIBUTING.md:415]: No need to edit the original post:

```suggestion
  comment.  If you are empowered to merge PRs, you should be empowered to [link
  the PR to the issue manually](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue-using-the-pull-request-sidebar), if necessary.
```

**Review by 1ucian0 — APPROVED** [review]
Great job. Some non-blocking comments, so approving. If any of the comments get committed, I can approve again.

**Comment by jakelishman** [comment]
I'd forgotten about Luciano's comments above, but just for the record: I'm going to merge this now, and Luciano, can you move your suggestions into a new PR and assign me?  Mostly I just want to get the main body merged to the main repo, rather than it staying in limbo.

**Review by Cryoris — APPROVED** [review]
